File: CPP/11. Container With Most Water.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
	int maxArea(vector<int>& height) {
		int i = 0;
		int j = height.size() - 1;
		int prod = 0;

		while (i < j) {
			prod = std::max<int>(prod, std::min<int>(height[i], height[j]) * (j - i));
			if (height[j] >= height[i]) {
				i++;
			}
			else {
				j--;
			}
		}

		return prod;
	}
};

int main() {
	Solution sol;
	vector<int> height = { 1, 8, 6, 2, 5, 4, 8, 3, 7 };
	cout << sol.maxArea(height) << endl;
}

File: Python/11. Container With Most Water.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        i = 0
        prod = 0
        j = len(height)-1
        while i<j:
            prod= max(prod, (min(height[i],height[j])*(j-i)))
            if height[j]>=height[i] :
                i+=1
            else:
                j-=1
        return prod
    
if __name__ == '__main__':
    height = [1,8,6,2,5,4,8,3,7]
    result = Solution().maxArea(height)
    print(result)